pkg/controller/endpointslice/utils.go

```diff
@@ -346,3 +350,44 @@ func (sl endpointSliceEndpointLen) Swap(i, j int) { sl[i], sl[j] = sl[j], sl[i]
 func (sl endpointSliceEndpointLen) Less(i, j int) bool {
 	return len(sl[i].Endpoints) > len(sl[j].Endpoints)
 }
+
+// returns a map of address types used by a service
+func getAddressTypesForService(service *corev1.Service) map[discovery.AddressType]struct{} {
+	serviceSupportedAddresses := make(map[discovery.AddressType]struct{})
+	// TODO: (khenidak) when address types are removed in favor of
+	// v1.IPFamily this will need to be removed, and work directly with
+	// v1.IPFamily types
+
+	// IMPORTANT: we assume that IP of (discovery.AddressType enum) is never in use
+	// as it gets deprecated
+	for _, family := range service.Spec.IPFamilies {
+		if family == corev1.IPv4Protocol {
+			serviceSupportedAddresses[discovery.AddressTypeIPv4] = struct{}{}
+		}
+
+		if family == corev1.IPv6Protocol {
+			serviceSupportedAddresses[discovery.AddressTypeIPv6] = struct{}{}
+		}
+	}
+
+	// Why do we need this:
+	// a cluster being upgraded to the new apis
+	// will have service.spec.IPFamilies: nil
+	// if the controller manager connected to old api
+	// server. This will have the nasty side effect of
+	// removing all slices already created for this service.
+	// this will disable all routing to service vip (ClusterIP)
+	// this ensures that this does not happen
+	// TODO (khenidak) remove when (1) dual stack becomes
+	// enabled by default (2) v1.19 falls of supported versions
+	if len(serviceSupportedAddresses) == 0 {
+		addrType := discovery.AddressTypeIPv4
+		if utilnet.IsIPv6String(service.Spec.ClusterIP) {
```

Comment:
headless service will be ipv4, right?

Reply:
you have comment here, https://github.com/kubernetes/kubernetes/pull/91824/commits/a47e33977f099c04728c544ec6050c6d123c564c#diff-cc9110ca78ea393b4d11154cce113bfbR290

maybe we should add another here too

Reply:
No. They depend on cluster config. However headless without selector can be anything.

Reply:
then, if I read this correctly, if `service.Spec.IPFamilies` is empty, and ClusterIP is None, addrType will be IPv4 (same as we have in the endpoints controller